src/SMap.cpp

```diff
@@ -65,67 +67,60 @@ SMap::SMap(const STable table)
             end = start + 1;
             while (end < 9 && !blank[end*9+j])
                 end++;
-            length = end - start;
-            Straight* str8t = new Straight(this, fields+start*9+j, fields+end*9+j, 9);
-            col_straights[j].push_back(str8t);
+            auto new_str = std::make_unique<Straight>(
+                this, & fields[start*9+j], & fields[end*9+j], 9, start*9 + j
+            );
+            col_straights[j].push_back(std::move(new_str));
             start = end + 1;
         }
         while (start < 9);
     }
 }
 
 SMap::SMap(const SMap& smp)
-{
+{    if (opt_debug) cerr << "SMap from SMap at " << this << " allocated\n";
+
     // violation of the rules sets this to true
     hash_value = smp.hash_value;
 
-    fields = new uint16_t[9*9];
-    blank = new bool[9*9];
-    fixed = new bool[9*9];
 
-    memcpy(&sure_cand_row, &smp.sure_cand_row, 9*sizeof(uint16_t));
-    memcpy(&sure_cand_col, &smp.sure_cand_col, 9*sizeof(uint16_t));
-    memcpy(fields, smp.fields, 81*sizeof(uint16_t));
-    memcpy(blank, smp.blank, 81*sizeof(bool));
-    memcpy(fixed, smp.fixed, 81*sizeof(bool));
+    sure_cand_row = smp.sure_cand_row;
+    sure_cand_col = smp.sure_cand_col;
+    fields = smp.fields;
+    blank = smp.blank;
+    fixed = smp.fixed;
 
     for (int i = 0; i < 9; i++) {
-        for (auto str8t : smp.row_straights[i]) {
-            Straight* str = new Straight(*str8t);
-            row_straights[i].push_back(str);
+        for (auto && old_str : smp.row_straights[i]) {
+            auto new_str = std::make_unique<Straight>(*old_str);
             // set pointers correctly
-            str->map = this;
-            str->start = str8t->start - smp.fields + fields;
-            str->end = str8t->end - smp.fields + fields;
+            new_str->map = this;
+            new_str->start = & fields[old_str->starts_at];
+            new_str->end = & fields[old_str->ends_at];
+            row_straights[i].push_back(std::move(new_str));
         }
-        for (auto str8t : smp.col_straights[i]) {
-            Straight* str = new Straight(*str8t);
-            col_straights[i].push_back(str);
+        for (auto && old_str : smp.col_straights[i]) {
+            auto new_str = std::make_unique<Straight>(*old_str);
             // set pointers correctly
-            str->map = this;
-            str->start = str8t->start - smp.fields + fields;
-            str->end = str8t->end - smp.fields + fields;
+            new_str->map = this;
+            new_str->start = & fields[old_str->starts_at];
+            new_str->end = & fields[old_str->ends_at];
+            col_straights[i].push_back(std::move(new_str));
         }
     }
 }
 
 SMap& SMap::operator=(SMap mp)
 {
-    SMap *new_map = new SMap(mp);
-    return *new_map;
+    assert(false);    // not used
```

Comment:
If it is not used, we can delete the operator. Instead of an error at runtime, we would get an error at compile time, which is preferable.

Reply:
I agree!